File: src/0149-Max-Points-on-a-Line/0149.py
class Solution:
        
    def maxPoints(self, points: List[List[int]]) -> int:
        def GCD(a, b):
            return a if b == 0 else GCD(b, a % b)

        res, n = 1, len(points)
        lines = defaultdict(int)
        for i in range(n):
            lines.clear()
            for j in range(i + 1, n):
                p1, p2 = points[i], points[j]
                a, b = p1[0] - p2[0], p1[1] - p2[1]
                c = GCD(a, b)
                a //= c
                b //= c
                lines[a * 20000 + b] += 1
            for _, v in lines.items():
                res = max(res, v + 1)
        return res

File: src/0149-Max-Points-on-a-Line/0149.cpp
class Solution {
 public:
  int maxPoints(vector<vector<int>>& points) {
    int res = 1, n = points.size();
    unordered_map<int, int> lines;
    for (int i = 0; i < n; i++) {
      lines.clear();
      for (int j = i + 1; j < n; j++) {
        auto& p1 = points[i];
        auto& p2 = points[j];

        auto a = p1[0] - p2[0];
        auto b = p1[1] - p2[1];
        auto c = GCD(a, b);
        a /= c, b /= c;
        lines[a * 20000 + b]++;
      }
      for (auto& [x, y] : lines) {
        res = max(res, y + 1);
      }
    }
    return res;
  }
  int GCD(int a, int b) { return b == 0 ? a : GCD(b, a % b); }
};